internal/registration/registration_test.go

```diff
@@ -0,0 +1,172 @@
+		dispatcherMock *registration.MockDispatcherClient
+		configManager  *configuration.Manager
+		hb             *heartbeat.Heartbeat
+		hw             = &hardware.Hardware{}
+		monitor        = &datatransfer.Monitor{}
+		os             = &os.OS{}
+		err            error
+	)
+
+	BeforeEach(func() {
+		datadir, err = ioutil.TempDir("", "registrationTest")
+		Expect(err).ToNot(HaveOccurred())
+
+		mockCtrl = gomock.NewController(GinkgoT())
+
+		wkwMock = workload.NewMockWorkloadWrapper(mockCtrl)
+		wkwMock.EXPECT().Init().Return(nil).AnyTimes()
+
+		dispatcherMock = registration.NewMockDispatcherClient(mockCtrl)
+
+		wkManager, err = workload.NewWorkloadManagerWithParams(datadir, wkwMock)
+		Expect(err).NotTo(HaveOccurred(), "Cannot start the Workload Manager")
+
+		configManager = configuration.NewConfigurationManager(datadir)
+
+		hb = heartbeat.NewHeartbeatService(dispatcherMock,
+			configManager,
+			wkManager,
+			hw,
+			monitor)
+
+	})
+
+	AfterEach(func() {
+		mockCtrl.Finish()
+		_ = osUtil.Remove(datadir)
+	})
+
+	Context("RegisterDevice", func() {
+
+		RegistrationMatcher := func() gomock.Matcher {
+			return regMatcher{}
+		}
+
+		It("Work as expected", func() {
+
+			// given
+			reg := registration.NewRegistration(
+				hw, os, dispatcherMock, configManager, hb, wkManager, monitor)
+
+			// then
+			dispatcherMock.EXPECT().Send(gomock.Any(), RegistrationMatcher()).Times(1)
+
+			//  when
+			reg.RegisterDevice()
+		})
+
+		It("Try to re-register", func() {
+
+			// given
+			reg := registration.NewRegistration(
+				hw, os, dispatcherMock, configManager, hb, wkManager, monitor)
+			reg.RetryAfter = 1
+
+			// then
+			dispatcherMock.EXPECT().Send(gomock.Any(), gomock.Any()).Return(
+				nil, fmt.Errorf("failed")).Times(1)
+			dispatcherMock.EXPECT().Send(gomock.Any(), RegistrationMatcher()).Times(1)
+
+			//  when
+			reg.RegisterDevice()
+			time.Sleep(2 * time.Second)
+		})
+
+	})
+
+	Context("Deregister", func() {
+		BeforeEach(func() {
+			err = ioutil.WriteFile(
+				fmt.Sprintf("%s/%s", datadir, deviceConfigName),
+				[]byte("{}"),
+				0777)
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("Works as expected", func() {
+
+			// given
+			reg := registration.NewRegistration(
+				hw, os, dispatcherMock, configManager, hb, wkManager, monitor)
+
+			wkwMock.EXPECT().List().AnyTimes()
+			wkwMock.EXPECT().RemoveTable().AnyTimes()
+			wkwMock.EXPECT().RemoveMappingFile().AnyTimes()
+
+			// when
+			err := reg.Deregister()
+
+			// then
+			Expect(err).NotTo(HaveOccurred())
+
+		})
+
+		It("Return error if anything fails", func() {
+
+			// given
+			reg := registration.NewRegistration(
+				hw, os, dispatcherMock, configManager, hb, wkManager, monitor)
+
+			wkwMock.EXPECT().List().Return(nil, fmt.Errorf("Failed"))
+			wkwMock.EXPECT().RemoveTable().AnyTimes()
+			wkwMock.EXPECT().RemoveMappingFile().AnyTimes()
+
+			// when
+			err := reg.Deregister()
+
+			// then
+			Expect(err).To(HaveOccurred())
+
+		})
+
+	})
+
```

Comment:
What do you think about adding test for register -> deregister -> register again?